### ISSUE TITLE
wLimit may decrease when quantile approaches 1

### ISSUE THREAD

**Issue by cyb70289** [tdunning/t-digest#152]
In my C++ implementation, tdigest buffer expands above compression (`delta`) size when inputs are very long. I find the tdigest buffer is flooded with centroids with `weight=1` at the end.
Below is my finding, would like to hear advice from @tdunning.

Code: https://github.com/tdunning/t-digest/blob/master/core/src/main/java/com/tdunning/math/stats/MergingDigest.java#L449
```
k1 = scale.k(wSoFar / totalWeight, normalizer);
wLimit = totalWeight * scale.q(k1 + 1, normalizer);
```
`wLimit` is the limit of accumulated weights for current centroid. I think it should be strictly increasing. But when quantile approaches 1, it actually decreases, which causes all latter inputs occupy one centroid of their own, so the long unit weight centroids at the end.
The reason why `wLimit` decreases is when quantile (`wSoFar / totalWeight`) approaches 1, `k1` approaches `delta / 2`, and `k1 + 1` exceeds `delta / 2`, out of the valid input [`-delta/2`, `delta/2`] of `q` function. E.g., for `sin/arcsin` K function, it means `k1 + 1` crosses the highest peak of `sin` curve and goes down.
My workaround is to compare new `wLimit` with last `wLimit`, if it decrease, set it to `totalWeight`. It does remove the long trailing unit weight centroids, and restrict the tdigest buffer size in bound.

**Comment by tdunning** [comment]
Hmm...

The q function should handle values of k outside the bounds gracefully.

Is it just K_1 that has this problem?  It should be easy to add a limit
comparison there.

I think that K_2 and K_3 are safe here. Both depend on exponentials which
behave gracefully no matter whether k exceeds reasonable bounds.

Do you have a specific test that demonstrates the problem (other than a
test on the scale function itself)?

**Comment by cyb70289** [comment]
Thanks. I think a limit comparison is okay.
I didn't investigate deep K_2 and K_3 functions.
I don't have specific test, just printed wLimit and find it decreases at the end.
Close this issue as it's no big issue.

**Comment by tdunning** [comment]
I am going to reopen this until I get the limit into place with tests.

**Comment by tdunning** [comment]
Closed with commit 07edffc2df4934ae0afbca6ef09c4674fd9ff30e
